### PR TITLE
feat(core): impl doc display meta extension

### PR DESCRIPTION
**Pull request by fundon**
Closes: [BS-2111](https://linear.app/affine-design/issue/BS-2111/定义和实现-docdisplaymetaextension)
Upstreams: https://github.com/toeverything/blocksuite/pull/8953

https://github.com/user-attachments/assets/008d7433-efef-47c4-8189-9bc288e61199

### CODE REVIEW

**Comment by fundon** [comment]
* **#9165** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9165?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9165?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=9165).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/root-block.ts [hunk 1]
```diff
@@ -141,6 +161,190 @@ function getThemeExtension(framework: FrameworkProvider) {
...
+      docId: string,
+      { params, title, referenced = false }: DocDisplayMetaParams = {}
+    ): Signal<TemplateResult> {
+      const doc = docsService.list.doc$(docId).value;
+
+      if (!doc) {
+        return signal(icons.deleted);
+      }
+
+      let icon = this.iconMap.get(doc);
```
Comment (EYHN) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/root-block.ts]: use DocDisplayMetaService

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/root-block.ts [hunk 2]
```diff
@@ -141,6 +161,190 @@ function getThemeExtension(framework: FrameworkProvider) {
...
+
+        return icon.value;
+      });
+    }
+
+    title(docId: string): Signal<string> {
+      const doc = docsService.list.doc$(docId).value;
+      if (!doc) return signal(I18n['com.affine.notFoundPage.title']());
+
+      let title = this.titleMap.get(doc);
```
Comment (EYHN) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/root-block.ts]: use docDisplayMetaService

Comment (fundon) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/custom/root-block.ts]: docId and links are a `1:N` relationship

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 3]
```diff
@@ -178,20 +181,33 @@ export class DocDisplayMetaService extends Service {
...
   title$(docId: string, options?: DocDisplayTitleOptions) {
     return LiveData.computed(get => {
+      const lng = get(this.i18nService.i18n.currentLanguageKey$);
       const doc = get(this.docsService.list.doc$(docId));
-      const docTitle = doc ? get(doc.title$) : undefined;
+      if (!doc) {
+        return this.i18nService.i18n.i18next.t(
+          'com.affine.notFoundPage.title',
+          { lng }
+        );
```
Comment (EYHN) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:190]:
```suggestion
        return I18n['com.affine.notFoundPage.title']()
```

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 4]
```diff
@@ -52,6 +53,7 @@ interface DocDisplayIconOptions<T extends IconType> {
 }
 interface DocDisplayTitleOptions {
   originalTitle?: string;
+  title?: string; // title alias
```
Comment (EYHN) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:56]: DocDisplayService no need to know alias.

alias should be handled at block level

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 5]
```diff
@@ -178,20 +181,33 @@ export class DocDisplayMetaService extends Service {
...
+      if (!doc) {
+        return this.i18nService.i18n.i18next.t(
+          'com.affine.notFoundPage.title',
+          { lng }
+        );
+      }
 
-      const journalDateString = get(this.journalService.journalDate$(docId));
+      // title alias
+      if (options?.title) return options.title;
```
Comment (EYHN) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts]: should handle emoji icon

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 6]
```diff
@@ -149,7 +152,7 @@ export class DocDisplayMetaService extends Service {
```
Comment (EYHN) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:152]: <i class='graphite__hidden'>[Re: line +145]<br/><br/></i>

should handle emoji icon<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9165?utm_source=unchanged-line-comment">Graphite</a>.</span>

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 7]
```diff
@@ -178,31 +181,48 @@ export class DocDisplayMetaService extends Service {
...
+      const docTitle = get(doc.title$);
+
       // empty title
-      if (!docTitle) return { i18nKey: 'Untitled' } as const;
+      if (!docTitle) {
+        return this.i18nService.i18n.i18next.t('Untitled', { lng });
+      }
 
       // reference
       if (options?.reference) return docTitle;
```
Comment (fundon) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:221]: Can delete it here?

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts [hunk 8]
```diff
@@ -178,31 +181,48 @@ export class DocDisplayMetaService extends Service {
...
-      const journalDateString = get(this.journalService.journalDate$(docId));
+      if (!doc) {
+        return this.i18nService.i18n.i18next.t(
+          'com.affine.notFoundPage.title',
+          { lng }
+        );
+      }
+
+      // title alias
+      if (options?.title) {
```
Comment (pengx17) [inline packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:191]: maybe alias should replace alias even if the target doc is not found?

**Review by pengx17 — APPROVED** [review]

File: packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts (hunk 8 above)
Comment (fundon) [reply packages/frontend/core/src/modules/doc-display-meta/services/doc-display-meta.ts:191]: good catch

**Comment by fundon** [comment]
### Merge activity

* **Dec 18, 8:01 PM EST**: The merge label 'merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE) once it meets the requirements.
* **Dec 18, 8:01 PM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).
* **Dec 18, 8:15 PM EST**: A user  merged this pull request with the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).

**Pull request merged**